### PR TITLE
feat: wallet service change server

### PR DESCRIPTION
**Pull request by andreabadesso**
## Video

https://user-images.githubusercontent.com/3586068/167901342-78e8ca9b-9c62-4ee4-9e3d-32cc4d27dc3f.mov

### Acceptance Criteria
- The user should be able to change the wallet service base url and websocket url
- The wallet should get the network from storage instead of getting it from the connected fullnode as we should not request the fullnode directly when using the wallet-service facade

### Security Checklist
- [X] Make sure you do not include new dependencies in the project unless strictly necessary and do not include dev-dependencies as production ones. More dependencies increase the possibility of one of them being hijacked and affecting us.

### CODE REVIEW

File: src/screens/Server.js [hunk 1]
```diff
@@ -87,26 +122,64 @@ class Server extends React.Component {
...
+      const versionData = await this.props.wallet.getVersionData();
+      if (versionData.network !== 'mainnet') {
+        const network = versionData.network;
+        let selectedNetwork = '';
+
+        // Network might be 'testnet-golf' or 'testnet-charlie'
+        if (network.startsWith('testnet')) {
+          selectedNetwork = 'testnet';
+        } else {
+          selectedNetwork = network;
```
Comment (tuliomir) [inline src/screens/Server.js:158]: Even though we prepare for values others than `^testnet*` and `mainnet`, I don't see clearly how they would be accepted on the rest of the process here.

Is this allowed now, or a future enhancement?

File: src/screens/Server.js [hunk 2]
```diff
@@ -199,20 +311,52 @@ class Server extends React.Component {
...
                 <option value=""> -- </option>
                 {renderServerOptions()}
               </select>
             </div>
           </div>
+          {
+            this.props.useWalletService && (
+              <div className="row mt-3">
+                <div className="col-12">
+                  { this.props.useWalletService && (
```
Comment (tuliomir) [inline src/screens/Server.js:328]: This line can only be rendered if `this.props.useWalletService` is true.
There should be no need to re-check it here.

File: src/utils/version.js [hunk 3]
```diff
@@ -26,27 +26,26 @@ const version = {
...
-          network = 'testnet';
-        }
-        helpers.updateNetwork(network);
-        resolve({...data, network});
-      }, (error) => {
-        reject();
-      });
-    });
-    return newPromise;
+  async checkApiVersion(wallet) {
```
Comment (tuliomir) [inline src/utils/version.js:29]: It would be important to describe this parameter on the JSDocs.

File: src/utils/version.js [hunk 4]
```diff
@@ -26,27 +26,26 @@ const version = {
...
+    store.dispatch(isVersionAllowedUpdate({
+      allowed: hathorLib.helpers.isVersionAllowed(data.version, hathorLib.constants.MIN_API_VERSION)
+    }));
+    // Set network in lib to use the correct address byte
+    let network;
+    if (data.network === 'mainnet') {
+      network = 'mainnet';
+    } else {
+      // Can we assume it will be testnet? I think it's safe
+      network = 'testnet';
```
Comment (tuliomir) [inline src/utils/version.js:42]: On `screens/Server.js` we are already preparing for networks other than `mainnet` and `testnet`.
Should we make any changes here as well? Or at least indicate how to refactor this on a future change?

File: src/utils/wallet.js [hunk 5]
```diff
@@ -316,6 +316,32 @@ const wallet = {
...
+   * @memberof Wallet
+   * @inner
+   **/
+  async handlePartialUpdate (wallet, updatedBalanceMap) {
+    const tokens = Object.keys(updatedBalanceMap);
+    const tokensHistory = {};
+    const tokensBalance = {};
+
+    for (const token of tokens) {
+      /* eslint-disable no-await-in-loop */
```
Comment (tuliomir) [inline src/utils/wallet.js:334]: I suggest disabling this restriction on the whole application.
`await` in loops are very useful!

**Review by tuliomir — CHANGES_REQUESTED** [review]

File: src/screens/Server.js [hunk 6]
```diff
@@ -68,17 +75,45 @@ class Server extends React.Component {
...
+      if (this.props.useWalletService && this.refs.newWsServer.value === '') {
+        invalidServer = true;
+        errorMessage = t`New real-time server is not valid`
+      }
+    } else {
+      if (this.state.selectedServer === '') {
+        invalidServer = true;
+        errorMessage = t`New server is not valid`
+      }
+      if (this.props.useWalletService && this.state.selectedWsServer === '') {
```
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/screens/Server.js:97]: They are checking different state parameters (`selectedServer` vs `selectedWsServer`) so we can't use `else if`

File: src/screens/Server.js [hunk 7]
```diff
@@ -68,17 +75,45 @@ class Server extends React.Component {
...
       return;
     }
-    let newServer = null;
+
+    let newBaseServer = null;
+    let newWsServer = null;
     if (this.state.newServer) {
-      newServer = this.refs.newServer.value;
+      newBaseServer = this.refs.newServer.value;
+      newWsServer = this.refs.newWsServer.value;
```
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/screens/Server.js]: I agree, done in e1a3d8d50b23ea65a82a0749cae519ad0a2a2655

File: src/screens/Server.js (hunk 2 above)
Comment (andreabadesso) [reply src/screens/Server.js]: Done in 563884155ba01329272a1ca1e60bf3110223da28

File: src/utils/version.js (hunk 3 above)
Comment (andreabadesso) [reply src/utils/version.js:30]: Done in 84545042b377c6827ae98f8defff3babb88aa0b1

File: src/utils/version.js [hunk 8]
```diff
@@ -26,27 +26,26 @@ const version = {
...
-      });
-    });
-    return newPromise;
+  async checkApiVersion(wallet) {
+    const data = await wallet.getVersionData();
+
+    // Update version allowed in redux
+    store.dispatch(isVersionAllowedUpdate({
+      allowed: hathorLib.helpers.isVersionAllowed(data.version, hathorLib.constants.MIN_API_VERSION)
+    }));
```
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/utils/version.js]: I agree, done in d09945b2d35aed8ffbb588b352c56d143954a2f1

What do you think?

File: src/utils/version.js (hunk 4 above)
Comment (andreabadesso) [reply src/utils/version.js]: I think we can accept `privatenet` here as well

Refactored in f95c758d204a7d5d9b78befad21794f25b2b52f7

What do you think? @tuliomir and @pedroferreira1

Comment (tuliomir) [reply src/utils/version.js]: For me it seems correct now: both files have the same rules for network naming.

File: src/screens/Server.js (hunk 1 above)
Comment (tuliomir) [reply src/screens/Server.js:161]: Solved on the `utils/version` file :+1:

File: src/screens/Server.js [hunk 9]
```diff
@@ -87,26 +125,63 @@ class Server extends React.Component {
...
+        this.setState({
+          selectedNetwork,
+        });
+
         // Go back to the previous server
         // If the user decides to continue with this change, we will update again
-        hathorLib.wallet.changeServer(currentServer);
+        this.props.wallet.changeServer(currentServer);
+        if (this.props.useWalletService) {
+          hathorLib.config.setWalletServiceBaseWsUrl(currentWsServer);
```
Comment (pedroferreira1) [inline src/screens/Server.js:171]: Why don't you use `this.props.wallet.changeWsServer(currentWsServer);`? The same flow used by the server

File: src/screens/Server.js [hunk 10]
```diff
@@ -87,26 +125,63 @@ class Server extends React.Component {
...
+        }
         $('#alertModal').modal('show');
-        this.setState({ loading: false });
```
Comment (pedroferreira1) [inline src/screens/Server.js:101]: Why did you remove this? The loading will continue showing with the modal?

File: src/screens/Server.js [hunk 11]
```diff
@@ -87,26 +125,63 @@ class Server extends React.Component {
...
+        // We are on mainnet, so set the network on the singleton and storage
+        hathorLib.config.setNetwork('mainnet');
+        helpers.updateNetwork('mainnet');
         this.executeServerChange();
       }
-    }, () => {
+    } catch (e) {
       // Go back to the previous server
-      hathorLib.wallet.changeServer(currentServer);
+      this.props.wallet.changeServer(currentServer);
```
Comment (pedroferreira1) [inline src/screens/Server.js:182]: You should go back to the ws server as well in case of wallet service, right?

File: src/screens/Server.js [hunk 12]
```diff
@@ -157,23 +241,65 @@ class Server extends React.Component {
...
+   * Update state of the selected base server
    *
    * @param {Object} e Event of select change
    */
-  handleSelectChange = (e) => {
-    this.setState({ selectedValue: DEFAULT_SERVERS[e.target.value] });
+  handleBaseURLSelectChange = (e) => {
+    if (e.target.value === '') {
+      return this.setState({
+        selectedWsServer: '',
```
Comment (pedroferreira1) [inline src/screens/Server.js:251]: Shouldn't this clean `selectedServer` instead of `selectedWsServer` state?

File: src/utils/wallet.js [hunk 13]
```diff
@@ -316,6 +315,32 @@ const wallet = {
...
+  /**
+   * Fetches both history and balance for the affected tokens in updatedBalanceMap
+   *
+   * @param {HathorWallet} wallet object
+   * @param {Object} updatedBalanceMap An object containing tokens as the keys and their updated balances as values
+   *
+   * @memberof Wallet
+   * @inner
+   **/
+  async handlePartialUpdate (wallet, updatedBalanceMap) {
```
Comment (pedroferreira1) [inline src/utils/wallet.js:327]: Is it new in this PR?

File: src/utils/wallet.js [hunk 14]
```diff
@@ -338,19 +363,20 @@ const wallet = {
...
+    // We are offline, the connection object is yet to be created
+    store.dispatch(isOnlineUpdate({ isOnline: false }));
+
     // Fetch metadata of all tokens registered
-    this.fetchTokensMetadata(dataToken.map((token) => token.uid), data.network);
+    this.fetchTokensMetadata(dataToken.map((token) => token.uid), network.name);
 
     const uniqueDeviceId = walletHelpers.getUniqueId();
-    const featureFlags = new FeatureFlags(uniqueDeviceId, data.network);
+    const featureFlags = new FeatureFlags(uniqueDeviceId, network.name);
```
Comment (pedroferreira1) [inline src/utils/wallet.js:379]: If I try to get a feature flag with network `privatenet`, then unleash will return that it does not exist and we will always use the old facade?

**Review by pedroferreira1 — CHANGES_REQUESTED** [review]

File: src/utils/wallet.js (hunk 14 above)
Comment (andreabadesso) [reply src/utils/wallet.js:379]: That is correct

File: src/screens/Settings.js [hunk 15]
```diff
@@ -22,7 +22,14 @@ import hathorLib from '@hathor/wallet-lib';
 import ModalAlertNotSupported from '../components/ModalAlertNotSupported';
 import { str2jsx } from '../utils/i18n';
 import version from '../utils/version';
+import { connect } from "react-redux";
 
+const mapStateToProps = (state) => {
+  return {
+    wallet: state.wallet,
```
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/screens/Settings.js]: I am not, removed it on 955a586dc7985a15c991c7a9ce801c571af51c38

File: src/utils/helpers.js [hunk 16]
```diff
@@ -57,6 +57,25 @@ const helpers = {
     currentWindow.reload();
   },
 
+  /**
+   * Reloads the page loaded on electron
+   *
+   * @memberof Helpers
+   * @inner
+   */
+  reloadElectron() {
```
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/utils/helpers.js:66]: I did exactly this on the mobile app, the idea is to make sure that every event listener from the current facade is cleared before changing to another facade

File: src/screens/Server.js [hunk 17]
```diff
@@ -49,7 +54,9 @@ class Server extends React.Component {
       selectedValue: '',
       loading: false,
       selectedServer: '',
+      selectedWsServer: '',
```
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/screens/Server.js:59]: Added on d5f10d139ecf6316ab57df2a448b21b91311ad12

File: src/screens/Server.js [hunk 18]
```diff
@@ -49,7 +54,9 @@ class Server extends React.Component {
...
       testnetError: '',
+      selectedNetwork: 'mainnet',
```
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/screens/Server.js]: For no particular reason, refactored it on d5f10d139ecf6316ab57df2a448b21b91311ad12

File: src/screens/Server.js (hunk 9 above)
Comment (andreabadesso) [reply src/screens/Server.js]: You're right, fixed in 3cf4d9b38afe21487362b75e9041eef061d29d09

File: src/screens/Server.js (hunk 11 above)
Comment (andreabadesso) [reply src/screens/Server.js:184]: I should, done in 2ae8b4214d1015beba3730f90580a3a0200f2494

File: src/screens/Server.js (hunk 12 above)
Comment (andreabadesso) [reply src/screens/Server.js]: Yes it should, thanks

Fixed in c41501ed53a74ef44d72fc34e64d15367cd17c67

File: src/utils/wallet.js (hunk 13 above)
Comment (andreabadesso) [reply src/utils/wallet.js]: It was not added on this PR, but for some reason this method was duplicated

Removed the duplicate on 016e5ec07686aaeb5007df5d4800b5420e2ed8aa

File: src/screens/Server.js (hunk 10 above)
Comment (andreabadesso) [reply src/screens/Server.js:101]: Not sure why I removed it, makes sense to stop the loading when we show the modal

Fixed in ce9279adb876c243c4294a16ae7c4617c05dbe03

File: src/utils/helpers.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (pedroferreira1) [reply src/utils/helpers.js:66]: Yeah, I know. But I thought you had done this in the PR you integrated with unleash. What changes in this PR required this method?

**Review by pedroferreira1 — APPROVED** [review]

File: src/utils/helpers.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (andreabadesso) [reply src/utils/helpers.js:66]: Apparently I messed up the git history while solving conflicts, causing this diff to think that this method has been added on this PR. It wasn't

No changes were done on this method on this PR

File: package-lock.json [hunk 19]
```diff
@@ -3159,19 +3172,66 @@
         "file-uri-to-path": "1.0.0"
       }
     },
+    "bip-schnorr": {
```
Comment (andreabadesso) [inline package-lock.json:3175]: This is a `bitcore-lib` dependency, was upgraded to `0.6.4` in a version greater than `8.25.10`

File: package-lock.json [hunk 20]
```diff
@@ -5211,6 +5271,15 @@
         "safer-buffer": "2.1.2"
       }
     },
+    "ecurve": {
```
Comment (andreabadesso) [inline package-lock.json:5274]: This is a dependency of `bip-schnorr`, which is a dependency of `bitcore-lib`, that is the dep tree:

```
│ ├─┬ bitcore-lib@8.25.28
│ │ ├── bech32@2.0.0
│ │ ├─┬ bip-schnorr@0.6.4
│ │ │ ├── bigi@1.4.2
│ │ │ ├─┬ ecurve@1.0.6
│ │ │ │ ├── bigi@1.4.2 deduped
│ │ │ │ └── safe-buffer@5.2.0 deduped
│ │ │ ├── js-sha256@0.9.0
│ │ │ ├─┬ randombytes@2.1.0
│ │ │ │ └── safe-buffer@5.2.0 deduped
│ │ │ └── safe-buffer@5.2.1
```

File: package-lock.json [hunk 21]
```diff
@@ -9912,6 +9981,11 @@
       "resolved": "https://registry.npmjs.org/js-base64/-/js-base64-2.5.2.tgz",
       "integrity": "sha512-Vg8czh0Q7sFBSUMWWArX/miJeBWYBPpdU/3M/DKSaekLMqrqVPaedp+5mZhie/r0lgrcaYBfwXatEew6gwgiQQ=="
     },
+    "js-sha256": {
```
Comment (andreabadesso) [inline package-lock.json:9984]: Also a dependency of `bitcore-lib`

```
│ ├─┬ bitcore-lib@8.25.28
│ │ ├── bech32@2.0.0
│ │ ├─┬ bip-schnorr@0.6.4
│ │ │ ├── bigi@1.4.2
│ │ │ ├─┬ ecurve@1.0.6
│ │ │ │ ├── bigi@1.4.2 deduped
│ │ │ │ └── safe-buffer@5.2.0 deduped
│ │ │ ├── js-sha256@0.9.0                          <---
│ │ │ ├─┬ randombytes@2.1.0
│ │ │ │ └── safe-buffer@5.2.0 deduped
│ │ │ └── safe-buffer@5.2.1
```

File: package-lock.json [hunk 22]
```diff
@@ -10703,6 +10777,21 @@
       "resolved": "https://registry.npmjs.org/merge-descriptors/-/merge-descriptors-1.0.1.tgz",
       "integrity": "sha1-sAqqVW3YtEVoFQ7J0blT8/kMu2E="
     },
+    "merge-options": {
```
Comment (andreabadesso) [inline package-lock.json:10780]: This is a sub-dependency of `unleash-proxy-client`:

```
├─┬ unleash-proxy-client@1.11.0
│ ├─┬ @react-native-async-storage/async-storage@1.17.4
│ │ └─┬ merge-options@3.0.4                                                         <---
│ │   └── is-plain-obj@2.1.0
│ ├── tiny-emitter@2.1.0
│ └── uuid@8.3.2
```

File: package-lock.json [hunk 23]
```diff
@@ -16440,6 +16529,11 @@
       "resolved": "https://registry.npmjs.org/timsort/-/timsort-0.3.0.tgz",
       "integrity": "sha1-QFQRqOfmM5/mTbmiNN4R3DHgK9Q="
     },
+    "tiny-emitter": {
```
Comment (andreabadesso) [inline package-lock.json:16532]: Also a dependency of `unleash-proxy-client`:

```
├─┬ unleash-proxy-client@1.11.0
│ ├─┬ @react-native-async-storage/async-storage@1.17.4
│ │ └─┬ merge-options@3.0.4
│ │   └── is-plain-obj@2.1.0
│ ├── tiny-emitter@2.1.0                                                             <---
│ └── uuid@8.3.2
```
